Allowed (valid) values UI

kmtracey opened the pull request:
## Summary
Update the data dictionary UI to allow the user to specify valid values for "select" properties. These are also included in the export and import of the data dictionary.

Spec doc is here: https://docs.google.com/document/d/1BPhdopyt07xNVv1uZ000094j25otKm6Deut1yXRSsyo/edit#heading=h.gvscgm6qaljn

## Feature Flag

Data dictionary

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Data dictionary has a UI that allows for defining expected values.
![Screenshot from 2021-08-25 09-26-42](https://user-images.githubusercontent.com/430048/130799375-c9e6b5fa-0fb9-4e4a-b1eb-49f159fe08ac.png)
![Screenshot from 2021-08-25 09-27-12](https://user-images.githubusercontent.com/430048/130799400-23a9e3ff-0563-461c-9f2c-f0843a5f6db2.png)


## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

I hope to add test coverage of the data dictonary export/import; this is not there yet.

### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-3436)

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

The migration in this PR is reversible and should be quick.

- [x] This PR can be reverted after deploy with no further considerations

kmtracey committed:
Beginnings of UI for specifying allowed values...

...for select data types in data dictionary.

- add model for storing the allowed values and their descriptions
- fetch existing allowed values when initial page data is fetched
- display allowed values and an edit button when appropriate
- next step: make edit button functional

kmtracey committed:
Address initial review comments: try to be a bit more idiomatic/concise, and move html into template where it belongs.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Use a uiElementKeyValueList to make the allowed values lists editable.

Additionally, change some text on the ui element (e.g. heading, add button) to be more user-friendly (reduce emphasis on "mapping"). Possibly these text bits should be made caller-configurable if these changes are not appropriate for all the places this widget is currently used.

Further, make the placeholders in the "mapping" fields caller-configurable.

Still TODO: enable "Save" button when the allowed values are edited, tests

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Enable Save button on allowed value modal "Done" click.

kmtracey committed:
Swith to the `on('change', ...)` made available via...

the `KeyValList` call to `hqMain.eventize` to enable the Save button, rather than listening to the Done button click.

This will fire regardless of how the modal is closed (x in top or Done button), which seems correct since the top X still records whatever changes have been made in what is subsequently shown to the user.

Also make the allowed values modal work properly for a new dynamically added property created after page load.

kmtracey committed:
Add the basics of export/import of allowed values in data dictionary.

kmtracey committed:
Address PR feedback and change to "valid values"...

...for user-visible references to what we have been calling allowed values. (Much of the code still uses the "alloewd value" terminology, if wanted I can work to change the internals as well, but that's a bigger task than time probably allows today.)

kmtracey committed:
Merge branch 'master' into allowed_values_ui

orangejenny commented:
@kmtracey Code all looks good. Could you add a screenshot or two to the PR description?

orangejenny commented:
@kmtracey Two more things to do because this adds a new SQL model:
1. To fix the current test failure, add the new model [here](https://github.com/dimagi/commcare-hq/blob/37cf083f26bb5bf3bc069818237322bf90ac5b80/corehq/apps/domain/deletion.py#L322-L324) and update [this test](https://github.com/dimagi/commcare-hq/blob/37cf083f26bb5bf3bc069818237322bf90ac5b80/corehq/apps/domain/tests/test_delete_domain.py#L572-L587). This will ensure the new model gets deleted when a domain is deleted - domains are stored in CouchDB, so we don't have real foreign key relationships between them and their associated SQL data.
2. Add the new model [here](https://github.com/dimagi/commcare-hq/blob/37cf083f26bb5bf3bc069818237322bf90ac5b80/corehq/apps/dump_reload/sql/dump.py#L138-L141), which is used to export entire domains to move them between servers.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Changes to account for new SQL model.

- include new model in domain export code
- ensure new model is deleted when domain is deleted

kmtracey committed:
Give user specific feedback on save of a too-long valid value. Refs QA-3469

kmtracey committed:
Beef up error catching/reporting on data dictionary upload/import.

Catch when a case type is missing valid values, and let the user know instead of producing a 500 error
Notice when a case property cell is empty, and let the user know
Notice when a property is listed as having valid values but not found in the base case type information, and let user know.

Also added some basic unit tests of the data dictionary export/import. Import tests for these error situations, export is a bit more basic (so far just tests the right sheets are in the exported file), but it should be easy enough to add on as the data dictionary is worked on more.

Refs QA-3457

orangejenny commented on corehq/apps/data_dictionary/tests/test_import_export.py:
```diff
@@ -0,0 +1,133 @@
+        self.client.login(username='test', password='foobar')
+
+    def test_clean_import(self):
+        fname = os.path.join(os.path.dirname(__file__), 'data', 'clean_data_dictionary.xlsx')
```
There's a util to do this: the test class can use the mixin `TestFileMixin` and that includes a `get_file` helper (based on a `file_path` attribute of the test class. Check out [WorkbookJSONReaderTest](https://github.com/dimagi/commcare-hq/blob/d127f07ceb000d6647a9003d46d337f35220a54f/corehq/util/tests/test_excel.py#L14) for an example.

orangejenny commented on corehq/apps/data_dictionary/views.py:
```diff
@@ -413,25 +423,43 @@ def _process_bulk_upload(bulk_file, domain):
+    for case_type in missing_valid_values:
+        errors.append(_('Missing valid values sheet for case type {}').format(case_type))
+
+    for case_type in allowed_value_info.keys():
```
Nitpick: this can just be `for case_type in allowed_value_info:`

orangejenny commented:
@kmtracey Tests look great! I made a few minor comments. Once those are addressed, this will be ready to merge. I switched the base branch so it won't merge into master - there's now a `jls/caktus-data-dictionary-release-2` branch for this and future PRs that'll be part of the same release.

kmtracey committed:
Couple of simplifications from review comments.

orangejenny commented:
There's one more qabug, but this is merging into a dev branch, not master, and we want to be able to open additional PRs against that branch, so I'm merging this.
